### PR TITLE
added tests for Book functionality

### PR DESCRIPTION
**Pull request by JeniaSan**
Знімок екрана 2023-09-25 о 15 20 41
Знімок екрана 2023-09-25 о 15 21 35

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -17,6 +17,7 @@
 		<java.version>17</java.version>
 		<lombok.mapstruct.binding.version>0.2.0</lombok.mapstruct.binding.version>
 		<mapstruct.version>1.5.5.Final</mapstruct.version>
+		<testcontainers.version>1.18.0</testcontainers.version>
```
Comment (Denis-Balako) [inline pom.xml]: Optional: 1.19.0 version available

File: src/test/java/com/polishuchenko/bookstore/controller/BookControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        kobzarDto.setDescription(kobzar.getDescription());
+        kobzarDto.setCoverImage(kobzar.getCoverImage());
+    }
+
+    @WithMockUser(username = "user")
+    @Test
+    @Sql(
+            scripts = {
+                    "classpath:database/books/add-book-kobzar-to-book-table.sql"
+            }, executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD
```
Comment (Denis-Balako) [inline src/test/java/com/polishuchenko/bookstore/controller/BookControllerTest.java]: If you have only 1 script, {} usage are redundant. Also, that's better to make a script string constants.

File: src/test/java/com/polishuchenko/bookstore/controller/BookControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    )
+    @Sql(
+            scripts = "classpath:database/books/delete-book-kobzar-from-book-table.sql",
+            executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD
+    )
+    @DisplayName("Get list of books")
+    void getAll_ReturnsBookDtosList() throws Exception {
```
Comment (Denis-Balako) [inline src/test/java/com/polishuchenko/bookstore/controller/BookControllerTest.java]:
```suggestion
    void getAll_validRequestDto_ReturnsBookDtosList() throws Exception {
```

File: src/test/java/com/polishuchenko/bookstore/controller/BookControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        List<BookDto> expected = List.of(kobzarDto);
+        MvcResult mvcResult = mockMvc.perform(get("/books")).andReturn();
```
Comment (Denis-Balako) [inline src/test/java/com/polishuchenko/bookstore/controller/BookControllerTest.java:106]: IDK if contentType() and andExpect() methods are important, so let mentor judge

File: src/test/java/com/polishuchenko/bookstore/controller/BookControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        kobzarDto.setId(kobzar.getId());
+        kobzarDto.setTitle(kobzar.getTitle());
+        kobzarDto.setAuthor(kobzar.getAuthor());
+        kobzarDto.setIsbn(kobzar.getIsbn());
+        kobzarDto.setPrice(kobzar.getPrice());
+        kobzarDto.setDescription(kobzar.getDescription());
+        kobzarDto.setCoverImage(kobzar.getCoverImage());
+    }
+
+    @WithMockUser(username = "user")
```
Comment (Denis-Balako) [inline src/test/java/com/polishuchenko/bookstore/controller/BookControllerTest.java]: Magic strings 🪄

File: src/test/java/com/polishuchenko/bookstore/controller/BookControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    @Sql(
+            scripts = {
+                    "classpath:database/books/add-book-kobzar-to-book-table.sql"
+            }, executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD
+    )
+    @Sql(
+            scripts = "classpath:database/books/delete-book-kobzar-from-book-table.sql",
+            executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD
+    )
+    @DisplayName("Get book by valid id")
```
Comment (Denis-Balako) [inline src/test/java/com/polishuchenko/bookstore/controller/BookControllerTest.java]:
```suggestion
    @DisplayName("Get book dto by valid id")
```

File: src/test/java/com/polishuchenko/bookstore/controller/BookControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+            scripts = {
+                    "classpath:database/books/add-book-kobzar-to-book-table.sql"
+            }, executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD
+    )
+    @Sql(
+            scripts = "classpath:database/books/delete-book-kobzar-from-book-table.sql",
+            executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD
+    )
+    @DisplayName("Delete book by valid id")
+    void delete_validId_ReturnsNothing() throws Exception {
```
Comment (Denis-Balako) [inline src/test/java/com/polishuchenko/bookstore/controller/BookControllerTest.java]: Optional:
```suggestion
    void delete_validId_successful() throws Exception {
```

File: src/test/java/com/polishuchenko/bookstore/service/BookServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    @DisplayName("Save book to DB")
+    public void save_validBook_ReturnsExpectedBook() {
+        when(bookMapper.toModel(kobzarRequestDto)).thenReturn(kobzar);
+        when(bookRepository.save(any(Book.class))).thenReturn(kobzar);
+        when(bookMapper.toDto(kobzar)).thenReturn(kobzarDto);
+
+        BookDto actual = bookService.save(kobzarRequestDto);
+
+        verify(bookRepository).save(kobzar);
+        assertNotNull(actual);
```
Comment (Denis-Balako) [inline src/test/java/com/polishuchenko/bookstore/service/BookServiceImplTest.java:108]: Does actual is equals to kobzarDto?

File: src/test/java/com/polishuchenko/bookstore/service/BookServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+    }
+
+    @Test
+    @DisplayName("Find all books by params")
+    public void getAllBooks_validParams_ReturnsBooksList() {
+        BookSearchParameters searchParameters = new BookSearchParameters(
+                new String[]{"kobzar"}, new String[0], new String[0], new String[0]);
+        Specification<Book> specification = mock(Specification.class);
+        List<Book> books = List.of(kobzar);
+        final List<BookDto> bookDtos = List.of(kobzarDto);
```
Comment (Denis-Balako) [inline src/test/java/com/polishuchenko/bookstore/service/BookServiceImplTest.java]: Why is it final?

File: src/test/java/com/polishuchenko/bookstore/service/CategoryServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        detectiveDto.setName(detective.getName());
+    }
+
+    @Test
+    @DisplayName("Find all categories")
+    public void findAll_ReturnsListWithAllCategories() {
+        when(categoryRepository.findAll()).thenReturn(List.of(detective));
+        when(categoryMapper.toDto(ArgumentMatchers.any())).thenReturn(detectiveDto);
+
+        List<CategoryDto> actual = categoryService.findAll();
```
Comment (Denis-Balako) [inline src/test/java/com/polishuchenko/bookstore/service/CategoryServiceImplTest.java:56]: If you implemented pagination to category service, provide pageble interface also

File: src/test/resources/database/books/add-book-kobzar-to-book-table.sql [hunk 11]
```diff
@@ -0,0 +1 @@
+insert into book(id, title, author, isbn, price, description, cover_image, is_deleted) VALUES (1, 'Kobzar', 'Taras Shevchenko', '978-3-16-148410-0', 50, 'some description', 'some link', false);
```
Comment (Denis-Balako) [inline src/test/resources/database/books/add-book-kobzar-to-book-table.sql]: Add for all your scripts an empty line in the end of the file

**Review by Denis-Balako — CHANGES_REQUESTED** [review]
Great job! Left some advice for your implementation.
And add screenshot with test coverage.

File: src/test/java/com/polishuchenko/bookstore/service/CategoryServiceImplTest.java (hunk 10 above)
Comment (JeniaSan) [reply src/test/java/com/polishuchenko/bookstore/service/CategoryServiceImplTest.java:56]: I didn't implement pagination in category service

File: src/test/java/com/polishuchenko/bookstore/service/BookServiceImplTest.java (hunk 8 above)
Comment (JeniaSan) [reply src/test/java/com/polishuchenko/bookstore/service/BookServiceImplTest.java:108]: Знімок екрана 2023-09-25 о 22 09 13

File: src/test/java/com/polishuchenko/bookstore/service/BookServiceImplTest.java (hunk 9 above)
Comment (JeniaSan) [reply src/test/java/com/polishuchenko/bookstore/service/BookServiceImplTest.java]: Знімок екрана 2023-09-25 о 22 09 13

File: src/test/java/com/polishuchenko/bookstore/controller/CategoryControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+        invalidRequestDto.setName("");
+    }
+
+    @WithMockUser(username = ADMIN_NAME, roles = {"ADMIN"})
+    @Test
+    @Sql(
+            scripts = DELETE_CATEGORY_SCRIPT, executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD
+    )
+    @DisplayName("Create new category and add to DB")
+    void create_validRequest_ReturnsCategoryDto() throws Exception {
```
Comment (vovasalyha) [inline src/test/java/com/polishuchenko/bookstore/controller/CategoryControllerTest.java:92]:
```suggestion
    void create_validRequest_returnsCategoryDto() throws Exception {
```
after `_` use lowercase. fix it in all places

**Review by vovasalyha — APPROVED** [review]
Everything seems fine, GJ. Just fix naming according to my comment

**Pull request merged**